server/activity_manager.go

```diff
@@ -27,241 +29,255 @@ func NewActivityManager(server *Server, activityService api.ActivityService) *Ac
+		case "DONE":
+			level = webhook.WebhookSuccess
+			title = "Issue resolved - " + meta.issue.Name
+		case "CANCELED":
+			title = "Issue canceled - " + meta.issue.Name
+		}
+	case api.ActivityIssueCommentCreate:
+		title = "Comment created"
+		link += fmt.Sprintf("#activity%d", activity.ID)
+	case api.ActivityIssueFieldUpdate:
+		update := new(api.ActivityIssueFieldUpdatePayload)
+		if err := json.Unmarshal([]byte(activity.Payload), update); err != nil {
+			m.s.l.Warn("Failed to post webhook event after changing the issue field, failed to unmarshal payload",
+				zap.String("issue_name", meta.issue.Name),
+				zap.Error(err))
+			return webhookCtx, err
+		}
+		switch update.FieldID {
+		case api.IssueFieldAssignee:
+			{
+				var oldAssignee, newAssignee *api.Principal
+				if update.OldValue != "" {
+					oldID, err := strconv.Atoi(update.OldValue)
+					if err != nil {
+						m.s.l.Warn("Failed to post webhook event after changing the issue assignee, old assignee id is not number",
+							zap.String("issue_name", meta.issue.Name),
+							zap.String("old_assignee_id", update.OldValue),
+							zap.Error(err))
+						return webhookCtx, err
 					}
+					principalFind := &api.PrincipalFind{
+						ID: &oldID,
+					}
+					oldAssignee, err = m.s.PrincipalService.FindPrincipal(ctx, principalFind)
+					if err != nil {
+						m.s.l.Warn("Failed to post webhook event after changing the issue assignee, failed to find old assignee",
+							zap.String("issue_name", meta.issue.Name),
+							zap.String("old_assignee_id", update.OldValue),
+							zap.Error(err))
+						return webhookCtx, err
+					}
+				}
 
-					metaList = append(metaList, webhook.WebhookMeta{
-						Name:  "Issue",
-						Value: meta.issue.Name,
-					})
-					metaList = append(metaList, webhook.WebhookMeta{
-						Name:  "Project",
-						Value: meta.issue.Project.Name,
-					})
-
-					err := webhook.Post(
-						hook.Type,
-						webhook.WebhookContext{
-							URL:          hook.URL,
-							Level:        level,
-							Title:        title,
-							Description:  create.Comment,
-							Link:         link,
-							CreatorName:  updater.Name,
-							CreatorEmail: updater.Email,
-							CreatedTs:    time.Now().Unix(),
-							MetaList:     metaList,
-						},
-					)
+				if update.NewValue != "" {
+					newID, err := strconv.Atoi(update.NewValue)
 					if err != nil {
-						// The external webhook endpoint might be invalid which is out of our code control, so we just emit a warning
-						m.s.l.Warn("Failed to post webhook event after changing the issue status",
-							zap.String("webhook_type", hook.Type),
-							zap.String("webhook_name", hook.Name),
+						m.s.l.Warn("Failed to post webhook event after changing the issue assignee, new assignee id is not number",
 							zap.String("issue_name", meta.issue.Name),
-							zap.String("status", string(meta.issue.Status)),
+							zap.String("old_assignee_id", update.NewValue),
```

Comment:
Looks like a bug in the old code. I will create a separate PR later to keep this PR just for refactoring purpose.

Reply:
LOL, it's easy to hide a bug in a very long function.